samples/azure/azure-service-fabric-routing/Core_6/Shared/ReceiverSideDistribution/Pipeline/DistributeMessagesBasedOnPayload.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Pipeline;
+
+namespace Shared
+{
+    class DistributeMessagesBasedOnPayload : IBehavior<IIncomingLogicalMessageContext, IIncomingLogicalMessageContext>
+    {
+        readonly string localPartitionKey;
+        readonly Forwarder forwarder;
+        readonly Func<object, string> mapper;
+
+        public DistributeMessagesBasedOnPayload(string localPartitionKey, Forwarder forwarder, Func<object, string> mapper)
+        {
+            this.localPartitionKey = localPartitionKey;
+            this.forwarder = forwarder;
+            this.mapper = mapper;
+        }
+
+        public Task Invoke(IIncomingLogicalMessageContext context, Func<IIncomingLogicalMessageContext, Task> next)
+        {
+            string messagePartitionKey;
+
+            var intent = GetMessageIntent(context);
+            var isSubscriptionMessage = intent == MessageIntentEnum.Subscribe || intent == MessageIntentEnum.Unsubscribe;
+            var isReply = intent == MessageIntentEnum.Reply;
+
+            if (isSubscriptionMessage || isReply)
+            {
+                return next(context);
+            }
+
+            if (!context.MessageHeaders.TryGetValue(PartitionHeaders.PartitionKey, out messagePartitionKey))
```

Comment:
I propose to change this behavior to something like 
```
 if(!context.MessageHeaders.Contains(PartitionHeaders.PartitionKey)
   {
           return next(context);
   }
   // all that we are doing now
```

Thanks to that we make the coupling between this and `DistributeMessagesBasedOnHeader` a bit more explicit. The header based behavior will pass the message through in two cases:
 1. The header is there and is correct. In such case we should just pass it through as well (this is the if part)
 2. The header is not there. This will be the rest.